public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts

```diff
@@ -808,4 +807,22 @@ export class AzureMonitorQueryCtrl extends QueryCtrl {
   toggleEditorMode() {
     this.target.appInsights.rawQuery = !this.target.appInsights.rawQuery;
   }
+
+  updateLocations() {
+    const { queryMode } = this.target.azureMonitor;
+    const locations = this.getLocations();
+    if (!this.target.azureMonitor.data[queryMode].locations.length) {
+      this.target.azureMonitor.data[queryMode].locations = locations;
```

Comment:
do we need to store data in the query model? I remember checking a while back and noticed that azure queries were HUGE in the saved dashboard json model as they contained a number of dropdown options. Like the persisted query model (target) was used as a cache to store options. Do you know if this is still true? Check panel json. 

Another problem I remember is that azure makes some of calls for each query (to fill dropdowns) when the query editor is loaded. Wonder if there are ways to cache that so not every query editor (for every row) needs to issue same request. But this could also be something that is fixed, was a couple of months ago since I noticed this. 

Main point question is is that this.target.azureMonitor.data is all those locations & resourceGroups used when issuing the query and need to be saved in the dashboard query model? or just there for the UI to show options?

Reply:
Not totally sure how it works in app insight and log analytics, but for multi resources queries in azure monitor the dropdown data is not stored on the target, only the selected values. Options are stored [here](https://github.com/grafana/grafana/blob/57776e48e45d4d6995aded17bab375878d912f44/public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts#L150..L151), and values [here](url). Resource group and location values needs to be an array since a multi select dropdown is being used.

Can look at how the json model is designed for app insight and log analytics. Maybe it's possible to optimize the way data is being stored in the json model there.